File: McpAppsDotnet/server/ViewStore.cs
using System.Collections.Concurrent;
using System.Text.Json;
using System.Text.Json.Nodes;
using ModelContextProtocol;
using ModelContextProtocol.Extensions.Apps;
using ModelContextProtocol.Protocol;

namespace McpAppsDotnet;

/// <summary>
/// Loads the Vite-bundled single-file HTML views and wraps them in the
/// <c>resources/read</c> result an MCP App host expects.
///
/// The build (see the <c>BuildUiViews</c> and <c>CopyUiViews</c> targets in the
/// .csproj) builds the Vite bundles and copies <c>ui/dist/stepN.html</c> into the
/// output <c>views/</c> folder, so the files sit next to the executable at
/// <see cref="AppContext.BaseDirectory"/>. Files are read once and cached — they
/// never change while the server runs.
/// </summary>
public sealed class ViewStore
{
    private static readonly string ViewsDir = Path.Combine(AppContext.BaseDirectory, "views");

    private readonly ConcurrentDictionary<string, string> _cache = new();

    /// <summary>Returns the bundled HTML for <paramref name="fileName"/> (e.g. <c>step1.html</c>).</summary>
    public string Html(string fileName) => _cache.GetOrAdd(fileName, Load);

    /// <summary>
    /// Builds the <c>resources/read</c> result for a UI resource: a single
    /// <see cref="TextResourceContents"/> carrying the HTML, the MCP Apps MIME
    /// type (<see cref="McpApps.HtmlMimeType"/>), and optional content-level
    /// <c>_meta.ui</c> (used by step 6 for CSP).
    /// </summary>
    /// <remarks>
    /// The Apps package ships helpers for the tool and the resource *listing*
    /// (<c>McpApps.SetAppUi</c> / <c>McpApps.SetResourceUi</c>) but none for the
    /// <em>content</em> level, because <see cref="ResourceContents.Meta"/> is a raw
    /// <see cref="JsonObject"/>. So we serialize the typed
    /// <see cref="McpUiResourceMeta"/> ourselves with
    /// <see cref="McpApps.SerializerOptions"/> — those options carry the
    /// source-generated contracts (and the camelCase/ignore-null naming) for the
    /// Apps types, so the result is exactly the JSON the host expects — and hang it
    /// under the <c>ui</c> key.
    /// </remarks>
    public ReadResourceResult Read(string uri, string fileName, McpUiResourceMeta? ui = null) => new()
    {
        Contents =
        [
            new TextResourceContents
            {
                Uri = uri,
                MimeType = McpApps.HtmlMimeType,
                Text = Html(fileName),
                Meta = ui is null ? null : MetaObject(ui),
            },
        ],
    };

    /// <summary>
    /// Serializes a <see cref="McpUiResourceMeta"/> into the raw <c>{ "ui": … }</c>
    /// <see cref="JsonObject"/> that the SDK's untyped <c>Meta</c> properties take.
    /// </summary>
    public static JsonObject MetaObject(McpUiResourceMeta ui) =>
        new() { ["ui"] = JsonSerializer.SerializeToNode(ui, McpApps.SerializerOptions) };

    private static string Load(string fileName)
    {
        var path = Path.Combine(ViewsDir, fileName);
        if (!File.Exists(path))
        {
            throw new McpException(
                $"View '{fileName}' not found at '{path}'. Build the UI first " +
                "(cd ../ui && npm install && npm run build), or run dotnet build without -p:SkipUiBuild=true.");
        }

        return File.ReadAllText(path);
    }
}

File: McpApps/src/server/lib/view-store.ts
import { readFile } from "node:fs/promises";
import path from "node:path";
import type { ReadResourceResult } from "@modelcontextprotocol/server";
import { APP_MIME } from "./ui-meta.js";

/**
 * Loads the Vite-bundled single-file HTML views (see `scripts/build-ui.ts`
 * and `src/ui/**`) and wraps them in the `resources/read` result shape an
 * MCP App host expects.
 *
 * Each view is built to `dist/ui/<step>/index.html` — one fully self
 * contained file (inlined CSS and JS, no external `src=`/`href=`) because
 * the host's CSP for a `ui://` resource starts at `default-src 'none'`.
 *
 * Caching is deliberately conditional. The .NET original (`ViewStore.cs`)
 * caches forever with a `ConcurrentDictionary`, which is fine for a
 * published binary but bit us during live demos of *this* sample: with
 * `npm run watch:ui` rebuilding a view in the background, a cached server
 * would keep serving the stale HTML until restarted, and "why doesn't my
 * change show up" is a bad thing to debug in front of an audience. So here
 * we only cache in production (`NODE_ENV=production`, i.e. the built,
 * deployed case where the files never change again); every other run
 * re-reads the file on every `resources/read`, trading a few milliseconds
 * for views that always reflect the latest build.
 */
export class ViewStore {
  private readonly cache = new Map<string, string>();

  /**
   * Builds the `resources/read` result for a UI resource.
   *
   * @param uri The resource URI as requested (usually the `ui://` URI the
   *   resource was registered under), echoed back into the content item.
   * @param step The step folder name under `dist/ui/`, e.g. `"step1-hello"`.
   * @param contentMeta Optional content-level `_meta` (see `ui-meta.ts`);
   *   step 6 uses this for its CSP allow-list.
   */
  async read(uri: string, step: string, contentMeta?: Record<string, unknown>): Promise<ReadResourceResult> {
    const html = await this.loadHtml(step);
    return {
      contents: [
        {
          uri,
          mimeType: APP_MIME,
          text: html,
          _meta: contentMeta,
        },
      ],
    };
  }

  private async loadHtml(step: string): Promise<string> {
    const cached = this.cache.get(step);
    if (cached !== undefined) {
      return cached;
    }

    // `import.meta.dirname` is this file's directory:
    // `McpApps/src/server/lib`. Three levels up is the repo's `McpApps/`
    // root, from which the build output lives at `dist/ui/<step>/index.html`.
    const filePath = path.resolve(import.meta.dirname, "../../../dist/ui", step, "index.html");

    let html: string;
    try {
      html = await readFile(filePath, "utf-8");
    } catch {
      throw new Error(`View "${step}" not built — run "npm run build:ui" first.`);
    }

    if (process.env.NODE_ENV === "production") {
      this.cache.set(step, html);
    }
    return html;
  }
}
